components/node-labeler/cmd/run.go

```diff
@@ -249,6 +262,150 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req reconcile.Request) (r
 	return reconcile.Result{}, nil
 }
 
+type WorkspaceCountController struct {
+	client.Client
+}
+
+func NewWorkspaceCountController(client client.Client) (*WorkspaceCountController, error) {
+	return &WorkspaceCountController{
+		Client: client,
+	}, nil
+}
+
+func (wc *WorkspaceCountController) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("workspace-count").
+		For(&workspacev1.Workspace{}).
+		WithEventFilter(workspaceFilter()).
+		Complete(wc)
+}
+
+func workspaceFilter() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			ws := e.Object.(*workspacev1.Workspace)
+			if ws.Status.Runtime == nil {
+				log.WithField("workspace", ws.Name).Info("workspace not ready yet")
+				return false
+			}
+
+			return ws.Status.Runtime != nil && ws.Status.Runtime.NodeName != ""
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			wsOld := e.ObjectOld.(*workspacev1.Workspace)
+			ws := e.ObjectNew.(*workspacev1.Workspace)
+			if wsOld.Status.Runtime == nil && ws.Status.Runtime != nil {
+				return true
+			}
+
+			// if we've seen runtime info before, there's no need to reconcile again
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			ws := e.Object.(*workspacev1.Workspace)
+			return ws.Status.Runtime != nil && ws.Status.Runtime.NodeName != ""
+		},
+	}
+}
+
+func (wc *WorkspaceCountController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log.WithField("request", req.NamespacedName.String()).Info("WorkspaceCountController reconciling")
+
+	var ws workspacev1.Workspace
+	if err := wc.Get(ctx, req.NamespacedName, &ws); err != nil {
+		if !errors.IsNotFound(err) {
+			log.WithError(err).WithField("workspace", req.NamespacedName).Error("unable to fetch Workspace")
+			return ctrl.Result{}, err
+		}
+		// If workspace not found, do a full reconciliation
+		log.WithField("workspace", req.NamespacedName).Info("Workspace not found, reconciling all nodes")
+		return wc.reconcileAllNodes(ctx)
+	}
+
+	if ws.Status.Runtime != nil && ws.Status.Runtime.NodeName != "" {
+		var workspaceList workspacev1.WorkspaceList
+		if err := wc.List(ctx, &workspaceList); err != nil {
+			log.WithError(err).Error("failed to list workspaces")
+			return ctrl.Result{}, err
+		}
+
+		count := 0
+		nodeName := ws.Status.Runtime.NodeName
+		for _, ws := range workspaceList.Items {
+			if ws.Status.Runtime != nil &&
+				ws.Status.Runtime.NodeName == nodeName &&
+				ws.DeletionTimestamp.IsZero() {
```

Comment:
How does the deletion timestamp matter here? 

What does it's value change before and after backups are done? 

To me, the presence of workspaces (regardless of deletion timestamp's value) is enough, because worst case, even if we miss an event, in 2m this controller will run again (and deleted workspaces might be gone). But, I am probably missing something. 😅

Reply:
This seems unnecessary indeed, let's get rid the check. It shouldn't matter if the workspace is stuck somehow before actually getting deleted, it's not the job of this controller to do more guesswork around the state of the system.

Thanks!